backend/app/services/query_service.py

```
@@ -0,0 +1,474 @@
+    session_id = session_id or str(uuid.uuid4())
+    router_model = getattr(settings, "QUERY_ROUTER_MODEL", "gpt-4o-mini")
+    synthesis_model = getattr(settings, "QUERY_SYNTHESIS_MODEL", "gpt-4o-mini")
+    entity_top_k = getattr(settings, "QUERY_ENTITY_TOP_K", 15)
+    community_top_k = getattr(settings, "QUERY_COMMUNITY_TOP_K", 10)
+    threshold = getattr(settings, "QUERY_SIMILARITY_THRESHOLD", 0.7)
+    ttl = getattr(settings, "CHAT_HISTORY_TTL_SECONDS", 86400)
+    history_window = getattr(settings, "CHAT_HISTORY_WINDOW", 6)
+    answer_cache_ttl = getattr(settings, "QUERY_ANSWER_CACHE_TTL", 3600)
+
+    # --- 0. Answer cache (skip pipeline for repeated identical questions) ---
+    question_hash = hashlib.md5(question.encode()).hexdigest()
```

Ruff (0.15.5)

[error] Probable use of insecure hash functions in `hashlib`: `md5`

(S324)